load a single layer via a query param

camallen opened the pull request:
Specify layer query param to load a specific layer from the API and then zoom.

camallen committed:
allow a single layer to load

layer query param loads a single approved layer from the API

camallen committed:
remove log statements

eatyourgreens commented on js/maps.js:
```diff
@@ -1,3 +1,5 @@
+import { queryParams } from './queryParams.js'
```
I've been avoiding `import` because it may not work in the browser.

eatyourgreens replied:
Ignore that, [MDN says it has good support now](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#Browser_compatibility).

camallen replied:
as the comment about good support on this import statement? It looked like it had ok support when i went this way - can remove if needed.

eatyourgreens commented on js/maps.js:
```diff
@@ -220,13 +209,32 @@ function zoomToFit() {
+API[getLayerFunc](eventName, layer)
   .then(buildLayersMenu)
-  .then(renderMapAndFit);
+  .then(renderMap)
```
I don't think `renderMap` is a promise, so I think the next `then` block will run immediately.

eatyourgreens replied:
Ignore that, [MDN says it has good support now](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#Browser_compatibility).

eatyourgreens commented on js/maps.js:
```diff
@@ -220,13 +209,32 @@ function zoomToFit() {
   .then(buildLayersMenu)
-  .then(renderMapAndFit);
+  .then(renderMap)
+  .then(zoomFunc);
```
What you want to do here, I think, is pass `zoomFunc` to Papaparse as the `complete` callback in the parser config. https://www.papaparse.com/docs#config

Maybe pass it as an argument to the `renderMap` function?

camallen replied:
I've done it slightly differently and promisified the `renderMap()` function and used the promise resolve as the callback to the parser config. 

I'm not sure if passing the callback function to renderMap is more optimal and clearer / cleaner in the code? Happy to change.

eatyourgreens replied:
Sounds good to me. `renderMap()` is set up to handle an array of layers, so you might need to use `Promise.all()`. I can never remember how iteration works for `nodeLists` returned from query selectors, which are slightly different from arrays.

camallen committed:
zoom the loaded layer image

promisify the renderMap function and use the complete callback from pappa parse config to  allow the map setup promise chain to resolve after the a csv layer data file is loaded.

camallen committed:
switch to passing zoomFunc as callback

remove the promise chaining, instead just pass the function callback to pappa parse  complete resolver.

eatyourgreens commented on js/maps.js:
```diff
@@ -179,32 +165,38 @@ function toggleLayer(event) {
-      } else {
-        readMapFile(url);
-      }
+  return new Promise(function (resolve, reject) {
```
If there's more than one selected checkbox, I think this will resolve for the first selected layer.

eatyourgreens commented on js/maps.js:
```diff
@@ -179,32 +165,38 @@ function toggleLayer(event) {
+        if (HEATMAPS[url]) {
+          const heatmapData = filteredMapData(HEATMAP_DATA[url]);
+          HEATMAPS[url].setData(heatmapData);
+          HEATMAPS[url].setMap(map);
```
If the layer is already cached in `HEATMAPS[url]`, then the promise doesn't resolve. Does that cause problems showing/hiding layers?

camallen committed:
avoid calling the fitEventBounds on each layer

only calculate the fit event bounts once.
